### PR TITLE
Add SmallNORB Dataset

### PR DESCRIPTION
**Pull request by mavanb**
I have added a wrapper for the smallNORB dataset. See https://cs.nyu.edu/~ylclab/data/norb-v1.0-small/.   

SmallNORB is a iseful intended for experiments in 3D object reocgnition from shape. It is, for example, used by Hinton et al. in Dynamic Routing Between Capsules (2017) and Matrix Capsules with EM routing (2018). 

Something that might be reconsidered: 

The data set contains stereo pair images, so each sample consists of a left and a right image. Therefore, I include 4 differents modes (is an instantion parameter) for this data set: 
- 'stereo': include both left and right. Return them as stereo pairs (left and right pair). Usage: in case the user wants to use the stereo effect. 
- 'all': include left and right. Return as single image (either left or right). Usage: in case the user wants include the whole data set, but wants to be able to shuffle. This mode thus allows
shuffling over pairs. 
- 'left': include only left. Return as single image (left). Usage: in case the user wants to exclude images that are very similiar. Only half the data set is loaded, which reduces memory usage. 
- 'right': same as 'left', but then with 'right' images.

### CODE REVIEW

**Comment by FancyCoder0** [comment]
Dear mavanb,

We are researchers working on identifying redundant development and duplicated pull requests. We have found there is a pull request: https://github.com/pytorch/vision/pull/493 which might be a potentially duplicate to this one. We would really appreciate if you could help us to validate and give us feedback.

Thank you very much for your time!

Sincerely,
Luyao

File: torchvision/datasets/smallnorb.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+                raise
+
+        # read train files
+        left_train_img, right_train_img = self._read_image_file(self.data_files["train"]["dat"]["name"])
+        train_info = self._read_info_file(self.data_files["train"]["info"]["name"])
+        train_label = self._read_label_file(self.data_files["train"]["cat"]["name"])
+
+        # read test files
+        left_test_img, right_test_img = self._read_image_file(self.data_files["test"]["dat"]["name"])
+        test_label = self._read_info_file(self.data_files["test"]["info"]["name"])
```
Comment (abhay-venkatesh) [inline torchvision/datasets/smallnorb.py]: should swap test_label and test_info here

File: torchvision/datasets/smallnorb.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        Args:
+            index (int): Index
+
+        Returns:
+            mode ``all'', ``left'', ``right'':
+                tuple: (image, target, info)
+            mode ``stereo'':
+                tuple: (image left, image right, target, info)
+        """
+        target = self.labels[index % 24300] if self.mode is "all" else self.labels[index]
```
Comment (Kelvin945) [inline torchvision/datasets/smallnorb.py:145]: 'is' need to change to "=="

**Comment by yassineAlouini** [comment]
Thanks @mavanb for the contribution and sorry for taking that long to get back at you.

As you might know, there is a new dataset API being designed and existing datasets are being ported as discussed [here](https://github.com/pytorch/vision/issues/5336).

There is also this issue that lists new datasets that the community wants to add: https://github.com/pytorch/vision/issues/3562   

For now, the focus is to finish porting existing datasets and then add new ones once the API is stable enough.

The best thing to do for now is to wait a bit until the new design is finalized and then get some help to port the dataset to the new design or do it yourself if you feel like it @mavanb. There is this [guide ](https://github.com/pytorch/vision/blob/main/torchvision/prototype/datasets/_builtin/README.md) that can be useful. Does that work for you?

Thanks again for you contribution and sorry for the delay.

**Comment by pmeier** [comment]
Given that this PR received a few comments from the community although it was never properly reviewed (sorry for that @mavanb!), I guess it is a good idea to add it to the list in #3562. @yassineAlouini could you take that up? Still, as mentioned above, we are revamping our dataset API and thus won't accept any new datasets until that is finished. Since the new API uses a new paradigm, we can't port the changes from this PR 1-to-1. Still, it can serve as basis so the work is not completely wasted.

**Comment by pmeier** [comment]
> Do you also want me to review the code? thinking

Nope, we can do that on the PR that ports this to the new API. You should however mention this PR.

**Comment by mavanb** [comment]
Sorry, I do not work with this dataset or even `pytorch` anymore. Feel free to close the PR and mention it elsewhere if someone wants to pick it up and reuse some elements.
